metadata/src/audio/item.rs

```diff
@@ -16,98 +18,290 @@ use librespot_core::{
-            restriction
-                .catalogue_strs
-                .iter()
-                .any(|restricted_catalogue| restricted_catalogue == user_catalogue)
-        }) {
-            if let Some(allowed_countries) = &premium_restriction.countries_allowed {
-                // A restriction will specify either a whitelast *or* a blacklist,
-                // but not both. So restrict availability if there is a whitelist
-                // and the country isn't on it.
-                if allowed_countries.iter().any(|allowed| country == allowed) {
-                    return Ok(());
-                } else {
-                    return Err(UnavailabilityReason::NotWhitelisted);
+                if track.duration <= 0 {
+                    return Err(Error::unavailable(MetadataError::InvalidDuration(
+                        track.duration,
+                    )));
                 }
-            }
 
-            if let Some(forbidden_countries) = &premium_restriction.countries_forbidden {
-                if forbidden_countries
+                if track.is_explicit {
+                    if let Some(value) = session.get_user_attribute("filter-explicit-content") {
+                        if &value == "1" {
+                            return Err(Error::unavailable(MetadataError::ExplicitContentFiltered));
+                        }
+                    }
+                }
+
+                let track_id = track.id;
+                let uri = track_id.to_uri()?;
+                let album = track.album.name;
+
+                let artists = track
+                    .artists
+                    .iter()
+                    .map(|a| a.name.clone())
+                    .collect::<Vec<String>>();
+
+                let album_artists = track
+                    .album
+                    .artists
                     .iter()
-                    .any(|forbidden| country == forbidden)
-                {
-                    return Err(UnavailabilityReason::Blacklisted);
+                    .map(|a| a.name.clone())
+                    .collect::<Vec<String>>();
+
+                let mut covers = track.album.covers;
+
+                covers.sort_by(|a, b| b.width.cmp(&a.width));
+
+                let covers = covers
+                    .iter()
+                    .filter_map(|cover| {
+                        let cover_id = cover.id.to_string();
+
+                        if !cover_id.is_empty() {
+                            let cover_image = CoverImage {
+                                url: image_url.replace("{file_id}", &cover_id),
+                                size: cover.size,
+                                width: cover.width,
+                                height: cover.height,
+                            };
+
+                            Some(cover_image)
+                        } else {
+                            None
+                        }
+                    })
+                    .collect();
```

Comment:
This could also be refactored into its own method (to avoid code duplication with the episode code).